scrypto/src/math/integer.rs

```diff
@@ -0,0 +1,745 @@
+(U64, U8, U64), (U64, U16, U64), (U64, U32, U64), (U64, U64, U64), (U64, U128, U128),
+(U64, U256, U256), (U64, U384, U384), (U64, U512, U512), (U64, I8, I64), (U64, I16, I64),
+(U64, I32, I64), (U64, I64, I64), (U64, I128, I128), (U64, I256, I256), (U64, I384, I384),
+(U64, I512, I512),
+
+(U128, u8, U128), (U128, u16, U128), (U128, u32, U128), (U128, u64, U128),
+(U128, u128, U128), (U128, i8, I128), (U128, i16, I128), (U128, i32, I128),
+(U128, i64, I128), (U128, i128, I128),
+(U128, U8, U128), (U128, U16, U128), (U128, U32, U128), (U128, U64, U128),
+(U128, U128, U128), (U128, U256, U256), (U128, U384, U384), (U128, U512, U512),
+(U128, I8, I128), (U128, I16, I128), (U128, I32, I128), (U128, I64, I128),
+(U128, I128, I128), (U128, I256, I256), (U128, I384, I384), (U128, I512, I512),
+
+(U256, u8, U256), (U256, u16, U256), (U256, u32, U256), (U256, u64, U256),
+(U256, u128, U256), (U256, i8, I256), (U256, i16, I256), (U256, i32, I256),
+(U256, i64, I256), (U256, i128, I256),
+(U256, U8, U256),  (U256, U16, U256), (U256, U32, U256), (U256, U64, U256),
+(U256, U128, U256), (U256, U256, U256), (U256, U384, U384), (U256, U512, U512),
+(U256, I8, I256), (U256, I16, I256), (U256, I32, I256), (U256, I64, I256),
+(U256, I128, I256), (U256, I256, I256), (U256, I384, I384), (U256, I512, I512),
+
+(U384, u8, U384), (U384, u16, U384), (U384, u32, U384), (U384, u64, U384),
+(U384, u128, U384), (U384, i8, I384), (U384, i16, I384), (U384, i32, I384),
+(U384, i64, I384), (U384, i128, I384),
+(U384, U8, U384), (U384, U16, U384), (U384, U32, U384), (U384, U64, U384),
+(U384, U128, U384), (U384, U256, U384), (U384, U384, U384), (U384, U512, U512),
+(U384, I8, I384), (U384, I16, I384), (U384, I32, I384), (U384, I64, I384),
+(U384, I128, I384), (U384, I256, I384), (U384, I384, I384), (U384, I512, I512),
+
+(U512, u8, U512), (U512, u16, U512), (U512, u32, U512), (U512, u64, U512),
+(U512, u128, U512), (U512, i8, I512), (U512, i16, I512), (U512, i32, I512),
+(U512, i64, I512), (U512, i128, I512),
+(U512, U8, U512), (U512, U16, U512), (U512, U32, U512), (U512, U64, U512),
+(U512, U128, U512), (U512, U256, U512), (U512, U384, U512), (U512, U512, U512),
+(U512, I8, I512), (U512, I16, I512), (U512, I32, I512), (U512, I64, I512),
+(U512, I128, I512), (U512, I256, I512), (U512, I384, I512), (U512, I512, I512)
+ }
+
+macro_rules! checked_impl_not_large {
+    ($($t:ident),*) => {
+        $(
+            impl Not for $t {
+                type Output = $t;
+
+                #[inline]
+                fn not(self) -> $t {
+                    self.0.iter().map(|x| x.not()).collect::<Vec<u8>>().try_into().unwrap()
+                }
+            }
+            forward_ref_unop! { impl Not, not for $t }
+        )*
+    }
+}
+
+macro_rules! checked_impl_not_small {
+    ($($t:ident),*) => {
+        $(
+            impl Not for $t {
+                type Output = $t;
+
+                #[inline]
+                fn not(self) -> $t {
+                    $t(!self.0)
```

Comment:
Is `!self.0` checked?

Reply:
It can't fail. Checked makes only sense with `add(), sub(), ...` basic math. With bit hacking there is not checked arithmetic.

Reply:
You're right. I was stupidly thinking as `-`.